File: src/main/java/com/martmists/ynet/network/Network.java
package com.martmists.ynet.network;

import com.martmists.ynet.YNetMod;
import com.martmists.ynet.api.BaseProvider;
import com.martmists.ynet.blockentities.ControllerBlockEntity;
import com.martmists.ynet.blocks.ControllerBlock;
import net.minecraft.block.Block;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.BlockView;

import java.util.*;

public class Network {
    public static Map<BlockPos, Network> networks = new HashMap<>();
    public static Map<Class<?>, Set<Class<? extends BaseProvider>>> tMap = new HashMap<>();
    public Set<BlockPos> cables;
    public Set<BlockPos> connectors;
    public BlockPos controller;

    public synchronized static void removeCable(BlockView world, BlockPos p) {
        for (Map.Entry<BlockPos, Network> e : networks.entrySet()) {
            Network n = e.getValue();
            if (n.cables.contains(p)) {
                n.reloadAllNodes(world);
            }
        }
    }

    public synchronized static void removeConnector(BlockView world, BlockPos p) {
        for (Map.Entry<BlockPos, Network> e : networks.entrySet()) {
            Network n = e.getValue();
            if (n.connectors.contains(p)) {
                n.reloadAllNodes(world);
            }
        }
    }

    public synchronized static void addCable(BlockView world, BlockPos p) {
        Set<Network> connectedNetworks = new HashSet<>();
        for (BlockPos pos : new BlockPos[]{p.up(), p.down(), p.north(), p.east(), p.south(), p.west()}) {
            for (Map.Entry<BlockPos, Network> e : networks.entrySet()) {
                Network n = e.getValue();
                if (n.connectors.contains(pos) || n.cables.contains(pos)) {
                    connectedNetworks.add(n);
                }
            }
        }
        for (Network n : connectedNetworks) {
            n.cables.add(p);
            Set<BlockPos> known = new HashSet<>();
            known.addAll(n.cables);
            known.addAll(n.connectors);
            getConnectedBlocks(world, p, known, n.cables, n.connectors);
        }
    }

    public synchronized static void addConnector(BlockView world, BlockPos p) {
        Set<Network> connectedNetworks = new HashSet<>();
        for (BlockPos pos : new BlockPos[]{p.up(), p.down(), p.north(), p.east(), p.south(), p.west()}) {
            Block b = world.getBlockState(pos).getBlock();
            if (b instanceof ControllerBlock) {
                connectedNetworks.add(((ControllerBlockEntity) world.getBlockEntity(pos)).network);
            } else {
                for (Map.Entry<BlockPos, Network> e : networks.entrySet()) {
                    Network n = e.getValue();
                    if (n.connectors.contains(pos) || n.cables.contains(pos)) {
                        connectedNetworks.add(n);
                    }
                }
            }
        }
        for (Network n : connectedNetworks) {
            n.connectors.add(p);
            Set<BlockPos> known = new HashSet<>();
            known.addAll(n.cables);
            known.addAll(n.connectors);
            getConnectedBlocks(world, p, known, n.cables, n.connectors);
        }
    }

    public static void getConnectedBlocks(BlockView world, BlockPos origin, Set<BlockPos> cables, Set<BlockPos> connectors) {
        getConnectedBlocks(world, origin, new HashSet<>(), cables, connectors);
    }

    public static void getConnectedBlocks(BlockView world, BlockPos origin, Set<BlockPos> exclude, Set<BlockPos> cables, Set<BlockPos> connectors) {
        ArrayDeque<BlockPos> toSearch = new ArrayDeque<>();
        toSearch.push(origin);
        exclude.add(origin);
        // Search origin first
        Block bo = world.getBlockState(origin).getBlock();
        if (bo == YNetMod.CABLE) {
            cables.add(origin);
        } else if (bo == YNetMod.CONNECTOR) {
            connectors.add(origin);
        }

        // Search connected
        while (!toSearch.isEmpty()) {
            BlockPos p = toSearch.removeFirst();
            for (BlockPos p2 : Arrays.asList(p.up(), p.down(), p.north(), p.south(), p.east(), p.west())) {
                if (exclude.contains(p2)) {
                    continue;
                }
                exclude.add(p2);
                Block b = world.getBlockState(p2).getBlock();
                if (b == YNetMod.CABLE) {
                    toSearch.add(p2);
                    cables.add(p2);
                } else if (b == YNetMod.CONNECTOR) {
                    toSearch.add(p2);
                    connectors.add(p2);
                }
            }
        }
    }

    private static void findProviderTypes(Class<?> cls, Set<Class<? extends BaseProvider>> ret) {
        tMap.putIfAbsent(cls, ret);
        if (cls != Object.class && cls != null) {
            if (BaseProvider.class.isAssignableFrom(cls)) {
                ret.add((Class<? extends BaseProvider>) cls);
            }
            findProviderTypes(cls.getSuperclass(), ret);
            for (Class<?> itf : cls.getInterfaces()) {
                findProviderTypes(itf, ret);
            }
        }
    }

    public void reloadAllNodes(BlockView world) {
        cables = new HashSet<>();
        connectors = new HashSet<>();
        ControllerBlockEntity be = ((ControllerBlockEntity) world.getBlockEntity(controller));
        getConnectedBlocks(be.getWorld(), controller, cables, connectors);
    }

    public void setController(BlockPos pos) {
        controller = pos;
        networks.put(pos, this);
    }

    public Set<BlockPos> getProviders(BlockView world) {
        return getProviders(world, BaseProvider.class);
    }

    public Set<BlockPos> getProviders(BlockView world, Class<BaseProvider> type) {
        Set<BlockPos> providers = new HashSet<>();
        connectors.forEach((c) -> {
            for (BlockPos p : Arrays.asList(c.up(), c.down(), c.north(), c.south(), c.east(), c.west())) {
                Block b = world.getBlockState(p).getBlock();
                Set<Class<? extends BaseProvider>> providerTypes = new HashSet<>();
                if (tMap.get(b.getClass()) == null) {
                    findProviderTypes(b.getClass(), providerTypes);
                }
                if (tMap.get(b.getClass()).contains(type) || (type == BaseProvider.class && b instanceof BaseProvider)) {
                    providers.add(p);
                }
            }
        });
        return providers;
    }
}

File: src/main/kotlin/com/martmists/ynet/network/Network.kt
package com.martmists.ynet.network

import com.martmists.ynet.YNetMod
import com.martmists.ynet.blockentity.ControllerBlockEntity
import com.martmists.ynet.util.YNetRegistry
import com.martmists.ynet.transfer.type.DisabledType
import com.martmists.ynet.transfer.type.Type
import net.minecraft.block.entity.BlockEntity
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World

class Network(val be: ControllerBlockEntity) {
    val world: World
        get() = be.world!!
    var didInit = false
    val cables = mutableSetOf<Long>()  // BlockPos.asLong
    val connectors = mutableSetOf<Long>()
    val channels = arrayOf<Channel<*>>(
            Channel.empty(this),
            Channel.empty(this),
            Channel.empty(this),
            Channel.empty(this),
            Channel.empty(this),
            Channel.empty(this),
            Channel.empty(this),
            Channel.empty(this),
            Channel.empty(this),
            Channel.empty(this)
    )

    init {
        networks.add(this)
    }

    fun rescan() {
        cables.clear()
        connectors.clear()
        scan(be.pos)
        val connected = getConnectedBlocks()

        for ((index, channel) in channels.withIndex()) {
            val new = Channel(this, mutableListOf(), channel.type)
            new.connectedBlocks.addAll(connected.map { itt ->
                val old = channel.connectedBlocks.firstOrNull { it.be == itt }
                val cbe = ConfiguredBlockEntity(this.be, itt, old?.priority ?: 0, old?.mode ?: InteractionMode.DISABLED, old?.filter ?: mutableListOf())
                cbe
            })
            channels[index] = new
        }
        be.markDirty()
    }

    private fun getNeighbors(pos: BlockPos) = listOf(pos.up(), pos.down(), pos.north(), pos.east(), pos.south(), pos.west()).map { it.asLong() }

    private fun scan(origin: BlockPos) {
        val scanned = mutableSetOf(origin.asLong())
        val toScan = mutableSetOf(*getNeighbors(origin).toTypedArray())
        while (toScan.isNotEmpty()) {
            val next = toScan.elementAt(0)
            toScan.remove(next)

            for (side in getNeighbors(BlockPos.fromLong(next))) {
                if (side !in scanned) {
                    scanned.add(side)

                    val b = world.getBlockState(BlockPos.fromLong(side)).block
                    if (b === YNetMod.CABLE) {
                        cables.add(side)
                    } else if (b === YNetMod.CONNECTOR) {
                        connectors.add(side)
                    } else {
                        continue
                    }
                    toScan.add(side)
                }
            }
        }
    }

    fun tick() {
        if (!didInit) {
            rescan()
            didInit = true
        }

        if (world.isClient) return

        channels.forEach {
            if (it.type != DisabledType) {
                println("ticking channel $it of type ${it.type}")
                YNetRegistry.getHandler(it.type)?.tick(it as Channel<Type>)
            }
        }
    }

    fun getConnectedBlocks(): List<BlockEntity> {
        if (connectors.isEmpty()) return listOf()

        val allBlocks = mutableSetOf<Long>()

        for (c in connectors) {
            val pos = BlockPos.fromLong(c)
            for (side in getNeighbors(pos)) {
                if (YNetMod.shouldConnect(world, BlockPos.fromLong(side))) {
                    allBlocks.add(side)
                }
            }
        }

        return allBlocks.mapNotNull {
            world.getBlockEntity(BlockPos.fromLong(it))
        }.toList()
    }

    companion object {
        val networks = mutableListOf<Network>()

        fun addConnector(world: World, pos: BlockPos) {
            for (side in listOf(pos.up(), pos.down(), pos.north(), pos.east(), pos.south(), pos.west())) {
                if (world.getBlockState(side).block === YNetMod.CONTROLLER) {
                    val be = (world.getBlockEntity(side) as ControllerBlockEntity)
                    be.network.connectors.add(pos.asLong())
                    be.network.rescan()
                }

                for (net in networks) {
                    if (side.asLong() in net.connectors || side.asLong() in net.cables) {
                        net.connectors.add(pos.asLong())
                        net.rescan()
                    }
                }
            }
        }

        fun removeConnector(pos: BlockPos) {
            for (net in networks) {
                if (pos.asLong() in net.connectors) {
                    net.rescan()
                }
            }
        }

        fun addCable(pos: BlockPos) {
            for (side in listOf(pos.up(), pos.down(), pos.north(), pos.east(), pos.south(), pos.west())) {
                for (net in networks) {
                    if (side.asLong() in net.connectors || side.asLong() in net.cables) {
                        net.cables.add(pos.asLong())
                        net.rescan()
                    }
                }
            }
        }

        fun removeCable(pos: BlockPos) {
            for (net in networks) {
                if (pos.asLong() in net.cables) {
                    net.rescan()
                }
            }
        }
    }
}